straxen/plugins/event_patternfit.py

```diff
@@ -0,0 +1,410 @@
+        dtype += strax.time_fields
+        return dtype
+    
+    def setup(self):
+        self.mean_sPhoton = self.config['MeanPEperPhoton']
+        
+        # Getting S1 AFT maps
+        self.s1_aft_map = straxen.InterpolatingMap(
+            straxen.get_resource(self.config['s1_aft_map'], fmt='json'))
+                    
+        # Getting optical maps
+        self.s1_pattern_map = straxen.InterpolatingMap(
+            straxen.get_resource(self.config['s1_optical_map'], fmt='pkl'))
+        self.s2_pattern_map = straxen.InterpolatingMap(
+            straxen.get_resource(self.config['s2_optical_map'], fmt='pkl'))
+        
+        # Getting gain model to get dead PMTs
+        self.to_pe = straxen.get_to_pe(self.run_id, self.config['gain_model'], self.config['n_tpc_pmts'])
+        self.dead_PMTs = np.where(self.to_pe==0)[0]
+        self.pmtbool = ~np.in1d(np.arange(0, self.config['n_tpc_pmts']), self.dead_PMTs)
+        self.pmtbool_top = self.pmtbool[:self.config['n_top_pmts']]
+        self.pmtbool_bottom = self.pmtbool[self.config['n_top_pmts']:self.config['n_tpc_pmts']]
+        
+    def compute(self, events):
+        
+        result = np.zeros(len(events), dtype=self.dtype)
+        result['time'], result['endtime'] = events['time'], strax.endtime(events)
+
+        # Computing LLH values for S1s
+        self.compute_s1_llhvalue(events, result)
+        
+        # Computing LLH values for S2s
+        self.compute_s2_llhvalue(events, result)
+        
+        # Computing binomial test for s1 area fraction top
+        positions = np.vstack([events['x'], events['y'], events['z']]).T
+        aft_prob = self.s1_aft_map(positions)
+        
+        result['s1_area_fraction_top_continuos_probability'] = [
+            s1_area_fraction_top_probability(_aft, _s1, _s1_aft) if not np.isnan(_aft*_s1*_s1_aft) 
+            else np.nan for _aft, _s1, _s1_aft in zip(aft_prob, events['s1_area'], events['s1_area_fraction_top'])]
+        
+        result['s1_area_fraction_top_discrete_probability'] = [
+            s1_area_fraction_top_probability(_aft, _s1, _s1_aft, mode='discrete') if not np.isnan(_aft*_s1*_s1_aft)
+            else np.nan for _aft, _s1, _s1_aft in zip(aft_prob, events['s1_area'], events['s1_area_fraction_top'])]
+        
+        result['s1_photon_fraction_top_continuos_probability'] = [
+            s1_area_fraction_top_probability(_aft, _s1/self.config['MeanPEperPhoton'], _s1_aft) if not np.isnan(_aft*_s1*_s1_aft) 
+            else np.nan for _aft, _s1, _s1_aft in zip(aft_prob, events['s1_area'], events['s1_area_fraction_top'])]
+        
+        result['s1_photon_fraction_top_discrete_probability'] = [
+            s1_area_fraction_top_probability(_aft, _s1/self.config['MeanPEperPhoton'], _s1_aft, mode='discrete') if not np.isnan(_aft*_s1*_s1_aft)
+            else np.nan for _aft, _s1, _s1_aft in zip(aft_prob, events['s1_area'], events['s1_area_fraction_top'])]
```

Comment:
Sorry I cannot follow this, can we break it down?

Reply:
Hi Joran, yes I try to simplify it. Let me tested before push it